crates/bindings-csharp/BSATN.Codegen/Utils.cs

```diff
@@ -9,6 +9,386 @@ namespace SpacetimeDB.Codegen;
 
 public static class Utils
 {
+    static readonly StringBuilder sb = new();
+
+    public static StringBuilder StringBuilder()
+    {
+        sb.Clear();
+        return sb;
+    }
+
+    public static readonly SymbolDisplayFormat fmt =
+        new(
+            globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Included,
+            typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+            memberOptions: SymbolDisplayMemberOptions.IncludeContainingType,
+            genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters
+        );
+
+    public static bool Always(SyntaxNode _sn, CancellationToken _ct) => true;
+
+    public static string Bool(bool x) => x ? "true" : "false";
+
+    public static ITypeSymbol Generic(ITypeSymbol sym, int idx)
+    {
+        if (sym is INamedTypeSymbol named)
+        {
+            return named.TypeArguments[idx];
+        }
+        else
+        {
+            throw new InvalidOperationException(
+                $"Failed to get generic type argument {idx} for type {sym.ToDisplayString()}"
+            );
+        }
+    }
+
+    public static void CollectType(HashSet<ITypeSymbol> syms, ITypeSymbol sym)
+    {
+        if (syms.Contains(sym))
+        {
+            return;
+        }
+
+        syms.Add(sym);
+
+        foreach (var f in sym.GetMembers().OfType<IFieldSymbol>())
+        {
+            CollectType(syms, f.Type);
+        }
+
+        if (sym.BaseType?.OriginalDefinition.ToString() == "SpacetimeDB.TaggedEnum<Variants>")
+        {
+            foreach (var f in GetSumElements(sym))
+            {
+                CollectType(syms, f.Type);
+            }
+        }
+
+        switch (sym)
+        {
+            case IArrayTypeSymbol array:
+                CollectType(syms, array.ElementType);
+                break;
+            case INamedTypeSymbol named:
+                foreach (var t in named.TypeArguments)
+                {
+                    CollectType(syms, t);
+                }
+                break;
+        }
+    }
+
+    public class AnalyzedType
+    {
+        public required ITypeSymbol sym;
+        public required int idx;
+        public required string fqn;
+        public required string alg;
+        public required string var;
+        public required string? prim;
+        public required TypeKind kind;
+        public required bool isOpt;
+        public required bool isInt;
+        public required bool isEq;
+    }
+
+    public enum TypeKind
+    {
+        Prim,
+        Enum,
+        Builtin,
+        Option,
+        Array,
+        List,
+        Map,
+        Sum,
+        Product,
+    }
+
+    public static Dictionary<ISymbol, AnalyzedType> AnalyzeTypes(HashSet<ITypeSymbol> syms)
```

Comment:
I think this one at least deserves some documentation; the idea is to process all types needed by table and reducer definitions at once (that includes all field types and all argument types), such that they only get instantiated once in the emitted output.

Initially, it was also driving the generation of AlgebraicType instances; but it's been decided to leave that for a later point in time and another PR. So this function still generates the info needed to emit table and reducer defs in the Module generator.

Reply:
This shouldn't be necessary; we already use structs for BSATN definitions, which are zero-cost to instantiate (they don't allocate or anything, they just use the default value on stack).

The only reason we store them in variables right now is for convenient aliasing in the codegen.